### PR TITLE
Add passive validation - dumping cloud-init logs

### PR DESCRIPTION
**Pull request by golubev-ivan**
Added dumping cloud-init logs if cloud-init is contained on the machine where the agent is placed.

Checked cases:
- On linux with installed cloud-init
- On linux without installed cloud-init
- On windows (without cloud-init)

### CODE REVIEW

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 1]
```diff
@@ -131,6 +139,46 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
             executionContext.Debug("Diagnostic file upload complete.");
         }
 
+        private async Task DumpCloudInitLogs(DateTime jobStartTimeUtc, string diagFolder)
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:142]: Please add /// comment

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 2]
```diff
@@ -105,6 +105,14 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
                 File.Copy(agentLogFile, destination);
             }
 
+            // Copy cloud-init log files from linux machines
+            if (PlatformUtil.RunningOnLinux)
+            {
+                executionContext.Debug("Dumping cloud-init logs.");
+                await DumpCloudInitLogs(jobStartTimeUtc, HostContext.GetDirectory(WellKnownDirectory.Diag));
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:112]: Please add fallback - so if there is an error, it won't crush agent, but will print an error message in logs.

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 3]
```diff
@@ -105,6 +105,14 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                executionContext.Debug("Dumping cloud-init logs is completed.");
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:113]: Will it be really completed if error happens? Probably some re-phrasing would make sense as well

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 4]
```diff
@@ -131,6 +139,46 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
             executionContext.Debug("Diagnostic file upload complete.");
         }
 
+        private async Task DumpCloudInitLogs(DateTime jobStartTimeUtc, string diagFolder)
+        {
+            var builder = new StringBuilder();
+
+            string cloudInit = WhichUtil.Which("cloud-init", trace: Trace);
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:146]: Do we need to check presence of cloud-init in Path environment variable (this will be done by WhichUtil.Which method)?

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 5]
```diff
@@ -131,6 +139,46 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                    fileName: cloudInit,
+                    arguments: arguments,
+                    environment: null,
+                    requireExitCodeZero: false,
+                    outputEncoding: null,
+                    killProcessOnCancel: false,
+                    cancellationToken: default(CancellationToken));
+            }
+
+            Trace.Info(builder.ToString());
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:179]: Please split the logic of this method up - this method searches for cloud-init, calls logs collection and trace retrieved data - could it make sense at least to move tracing out of this method?

File: src/Agent.Worker/DiagnosticLogManager.cs (hunk 4 above)
Comment (golubev-ivan) [reply src/Agent.Worker/DiagnosticLogManager.cs:146]: As far as I see `WhichUtil.Which` is the safe way to check availability of cloud-init and to find out the full path to it. Or not recommend using this tool?

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 6]
```diff
@@ -131,6 +142,55 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+        /// Dumping cloud-init logs to diag folder of agent if cloud-init is installed on current machine.
+        /// </summary>
+        /// <param name="jobStartTimeUtc">Job start time</param>
+        /// <param name="diagFolder">Path to agent diag folder</param>
+        /// <returns>Returns the method execution logs</returns>
+        private async Task<string> DumpCloudInitLogs(DateTime jobStartTimeUtc, string diagFolder)
+        {
+            var builder = new StringBuilder();
+            string cloudInit = WhichUtil.Which("cloud-init", trace: Trace);
+            if (string.IsNullOrEmpty(cloudInit)) return "Cloud-init ins't found on current machine.";
```
Comment (EzzhevNikita) [inline src/Agent.Worker/DiagnosticLogManager.cs:155]: nit: `isn't`

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 7]
```diff
@@ -131,6 +142,55 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                {
+                    processInvoker.OutputDataReceived += (object sender, ProcessDataReceivedEventArgs args) =>
+                    {
+                        builder.AppendLine(args.Data);
+                    };
+
+                    processInvoker.ErrorDataReceived += (object sender, ProcessDataReceivedEventArgs args) =>
+                    {
+                        builder.AppendLine(args.Data);
+
```
Comment (EzzhevNikita) [inline src/Agent.Worker/DiagnosticLogManager.cs:172]: please remove extra line

**Review by EzzhevNikita — APPROVED** [review]

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 8]
```diff
@@ -131,6 +142,54 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+        /// Dumping cloud-init logs to diag folder of agent if cloud-init is installed on current machine.
+        /// </summary>
+        /// <param name="jobStartTimeUtc">Job start time</param>
+        /// <param name="diagFolder">Path to agent diag folder</param>
+        /// <returns>Returns the method execution logs</returns>
+        private async Task<string> DumpCloudInitLogs(DateTime jobStartTimeUtc, string diagFolder)
+        {
+            var builder = new StringBuilder();
+            string cloudInit = WhichUtil.Which("cloud-init", trace: Trace);
+            if (string.IsNullOrEmpty(cloudInit)) return "Cloud-init isn't found on current machine.";
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:155]: Nit: please avoid one line if operator

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks! Please check current comments also

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft Corporation.
+﻿// Copyright (c) Microsoft Corporation.
```
Comment (DaniilShmelev) [inline src/Agent.Worker/DiagnosticLogManager.cs:1]: Looks like you've saved this file in a different encoding - UTF-8 with BOM. This file was UTF-8 before. Could you please revert this change?

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 10]
```diff
@@ -131,6 +142,54 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                }
+            }
+            catch (Exception ex)
+            {
+                builder.AppendLine(ex.Message);
+            }
+            return builder.ToString();
+        }
+
+
```
Comment (DaniilShmelev) [inline src/Agent.Worker/DiagnosticLogManager.cs:192]:
```suggestion

```

**Review by DaniilShmelev — APPROVED** [review]
